// @makakwastaken/ts-edifact/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.VATDEC = exports.TAXCON = exports.PARTIN = exports.OSTRPT = exports.OSTENQ = exports.ORDERS = exports.INVRPT = exports.INVOIC = exports.IFTMIN = exports.GENRAL = exports.DESADV = exports.BALANC = exports.AUTHOR = exports.APERAK = exports.ValidatorStates = exports.ValidatorImpl = exports.NullValidator = exports.Dictionary = exports.Tracker = exports.Pointer = exports.Tokenizer = exports.SegmentTableBuilder = exports.Reader = exports.Parser = exports.Message = exports.InterchangeBuilder = exports.Group = exports.Edifact = exports.sanitizeFloat = exports.TradacomsSeparatorsBuilder = exports.Separators = exports.EdifactSeparatorsBuilder = exports.AnsiX12SeparatorsBuilder = exports.UNECEMessageStructureParser = exports.Configuration = exports.Cache = void 0;
var cache_1 = require("./cache");
Object.defineProperty(exports, "Cache", { enumerable: true, get: function () { return cache_1.Cache; } });
var configuration_1 = require("./configuration");
Object.defineProperty(exports, "Configuration", { enumerable: true, get: function () { return configuration_1.Configuration; } });
var messageStructureParser_1 = require("./edi/messageStructureParser");
Object.defineProperty(exports, "UNECEMessageStructureParser", { enumerable: true, get: function () { return messageStructureParser_1.UNECEMessageStructureParser; } });
var separators_1 = require("./edi/separators");
Object.defineProperty(exports, "AnsiX12SeparatorsBuilder", { enumerable: true, get: function () { return separators_1.AnsiX12SeparatorsBuilder; } });
Object.defineProperty(exports, "EdifactSeparatorsBuilder", { enumerable: true, get: function () { return separators_1.EdifactSeparatorsBuilder; } });
Object.defineProperty(exports, "Separators", { enumerable: true, get: function () { return separators_1.Separators; } });
Object.defineProperty(exports, "TradacomsSeparatorsBuilder", { enumerable: true, get: function () { return separators_1.TradacomsSeparatorsBuilder; } });
var edifact_1 = require("./edifact");
Object.defineProperty(exports, "sanitizeFloat", { enumerable: true, get: function () { return edifact_1.sanitizeFloat; } });
var interchangeBuilder_1 = require("./interchangeBuilder");
Object.defineProperty(exports, "Edifact", { enumerable: true, get: function () { return interchangeBuilder_1.Edifact; } });
Object.defineProperty(exports, "Group", { enumerable: true, get: function () { return interchangeBuilder_1.Group; } });
Object.defineProperty(exports, "InterchangeBuilder", { enumerable: true, get: function () { return interchangeBuilder_1.InterchangeBuilder; } });
Object.defineProperty(exports, "Message", { enumerable: true, get: function () { return interchangeBuilder_1.Message; } });
var parser_1 = require("./parser");
Object.defineProperty(exports, "Parser", { enumerable: true, get: function () { return parser_1.Parser; } });
var reader_1 = require("./reader");
Object.defineProperty(exports, "Reader", { enumerable: true, get: function () { return reader_1.Reader; } });
var segments_1 = require("./segments");
Object.defineProperty(exports, "SegmentTableBuilder", { enumerable: true, get: function () { return segments_1.SegmentTableBuilder; } });
var tokenizer_1 = require("./tokenizer");
Object.defineProperty(exports, "Tokenizer", { enumerable: true, get: function () { return tokenizer_1.Tokenizer; } });
var tracker_1 = require("./tracker");
Object.defineProperty(exports, "Pointer", { enumerable: true, get: function () { return tracker_1.Pointer; } });
Object.defineProperty(exports, "Tracker", { enumerable: true, get: function () { return tracker_1.Tracker; } });
var validator_1 = require("./validator");
Object.defineProperty(exports, "Dictionary", { enumerable: true, get: function () { return validator_1.Dictionary; } });
Object.defineProperty(exports, "NullValidator", { enumerable: true, get: function () { return validator_1.NullValidator; } });
Object.defineProperty(exports, "ValidatorImpl", { enumerable: true, get: function () { return validator_1.ValidatorImpl; } });
Object.defineProperty(exports, "ValidatorStates", { enumerable: true, get: function () { return validator_1.ValidatorStates; } });
const APERAK = __importStar(require("./messageSpec/APERAK.struct.json"));
exports.APERAK = APERAK;
const AUTHOR = __importStar(require("./messageSpec/AUTHOR.struct.json"));
exports.AUTHOR = AUTHOR;
const BALANC = __importStar(require("./messageSpec/BALANC.struct.json"));
exports.BALANC = BALANC;
const DESADV = __importStar(require("./messageSpec/DESADV.struct.json"));
exports.DESADV = DESADV;
const GENRAL = __importStar(require("./messageSpec/GENRAL.struct.json"));
exports.GENRAL = GENRAL;
const IFTMIN = __importStar(require("./messageSpec/IFTMIN.struct.json"));
exports.IFTMIN = IFTMIN;
const INVOIC = __importStar(require("./messageSpec/INVOIC.struct.json"));
exports.INVOIC = INVOIC;
const INVRPT = __importStar(require("./messageSpec/INVRPT.struct.json"));
exports.INVRPT = INVRPT;
const ORDERS = __importStar(require("./messageSpec/ORDERS.struct.json"));
exports.ORDERS = ORDERS;
const OSTENQ = __importStar(require("./messageSpec/OSTENQ.struct.json"));
exports.OSTENQ = OSTENQ;
const OSTRPT = __importStar(require("./messageSpec/OSTRPT.struct.json"));
exports.OSTRPT = OSTRPT;
const PARTIN = __importStar(require("./messageSpec/PARTIN.struct.json"));
exports.PARTIN = PARTIN;
const TAXCON = __importStar(require("./messageSpec/TAXCON.struct.json"));
exports.TAXCON = TAXCON;
const VATDEC = __importStar(require("./messageSpec/VATDEC.struct.json"));
exports.VATDEC = VATDEC;
